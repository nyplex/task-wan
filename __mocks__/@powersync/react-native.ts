module.exports = {
  PowerSyncDatabase: jest.fn().mockImplementation(() => ({
    connect: jest.fn(),
    ready: true,
  })),
};
