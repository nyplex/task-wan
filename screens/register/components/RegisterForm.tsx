import { useForm, Controller } from "react-hook-form";
import { useSelector } from "react-redux";
import { selectAuthStatus } from "@/redux/slices/authSlice/authSelectors";
import { VStack } from "@/components/gluestack/vstack";
import { Box } from "@/components/gluestack/box";
import { isValidEmail } from "@/utils/isValidEmail";
import useAuth from "@/hooks/useAuth";
import Button from "@/components/buttons/Button";
import Input from "@/components/form/Input";

const RegisterForm = () => {
  const { signup } = useAuth();
  const isLoading = useSelector(selectAuthStatus);

  const {
    control,
    handleSubmit,
    formState: { errors, isSubmitting },
  } = useForm<{
    email: string;
    username: string;
  }>({
    mode: "onSubmit",
    reValidateMode: "onChange",
    defaultValues: {
      email: "",
      username: "",
    },
  });

  const onSubmit = handleSubmit(async (data) => {
    await signup(data.email, data.username);
  });

  return (
    <VStack className="gap-2 mt-4">
      <Controller
        control={control}
        rules={{
          required: {
            value: true,
            message: "Email is required",
          },
          validate: (value) => isValidEmail(value.trim()) || "Invalid email",
        }}
        render={({ field: { onChange, onBlur, value }, fieldState: { error } }) => (
          <Input
            keyboardType="email-address"
            leftIcon="mail"
            value={value}
            autoCorrect={false}
            onChangeText={(text) => onChange(text)}
            onBlur={onBlur}
            isInvalid={!!error}
            isDisabled={isLoading}
            placeholder="Email"
            invalidText={error?.message || " "}
          />
        )}
        name="email"
      />
      <Controller
        control={control}
        rules={{
          required: {
            value: true,
            message: "Email is required",
          },
          maxLength: {
            value: 25,
            message: "Username must be 25 characters or less",
          },
          minLength: {
            value: 2,
            message: "Username must be at least 2 characters",
          },
        }}
        render={({ field: { onChange, onBlur, value } }) => (
          <Input
            keyboardType="default"
            leftIcon="user"
            value={value}
            autoCorrect={false}
            onChangeText={(text) => onChange(text.trim().replace(/\s+/g, "").toLowerCase())}
            onBlur={onBlur}
            isInvalid={!!errors.username}
            isDisabled={isLoading}
            placeholder="Username"
            invalidText={errors.username?.message || " "}
          />
        )}
        name="username"
      />
      <Box className="mt-8">
        <Button
          title="Register"
          isLoading={isLoading || isSubmitting}
          onPress={onSubmit}
          disabled={isLoading || isSubmitting}
        />
      </Box>
    </VStack>
  );
};

export default RegisterForm;
