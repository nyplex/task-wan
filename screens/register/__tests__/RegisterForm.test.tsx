import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import RegisterForm from "../components/RegisterForm";

let mockSignup: jest.Mock;
jest.mock("@/hooks/useAuth", () => ({
  __esModule: true,
  default: () => ({
    get signup() {
      return mockSignup;
    },
  }),
}));
jest.mock("react-redux", () => ({ useSelector: () => false }));

describe("RegisterForm", () => {
  it("renders inputs and button", () => {
    const { getByPlaceholderText, getByText } = render(<RegisterForm />);
    expect(getByPlaceholderText("Email")).toBeTruthy();
    expect(getByPlaceholderText("Username")).toBeTruthy();
    expect(getByText("Register")).toBeTruthy();
  });

  it("matches the snapshot", () => {
    const { toJSON } = render(<RegisterForm />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("disables register button and does not call signup for invalid email", async () => {
    mockSignup = jest.fn();
    const { getByPlaceholderText, getByText } = render(<RegisterForm />);
    fireEvent.changeText(getByPlaceholderText("Email"), "invalid-email");
    fireEvent.changeText(getByPlaceholderText("Username"), "validuser");
    const button = getByText("Register");
    fireEvent.press(button);
    expect(mockSignup).not.toHaveBeenCalled();
  });

  it("disables register button and does not call signup for invalid username", async () => {
    mockSignup = jest.fn();
    const { getByPlaceholderText, getByText } = render(<RegisterForm />);
    fireEvent.changeText(getByPlaceholderText("Email"), "test@example.com");
    fireEvent.changeText(getByPlaceholderText("Username"), "a"); // too short
    const button = getByText("Register");
    fireEvent.press(button);
    expect(mockSignup).not.toHaveBeenCalled();
  });

  it("calls signup on valid submit", async () => {
    mockSignup = jest.fn();
    const { getByPlaceholderText, getByText } = render(<RegisterForm />);
    fireEvent.changeText(getByPlaceholderText("Email"), "test@example.com");
    fireEvent.changeText(getByPlaceholderText("Username"), "validuser");

    // Traverse up to the pressable button
    let button = getByText("Register");
    while (button && !button.props.onPress) {
      button = button.parent;
    }

    await waitFor(() => {
      expect(button.props.disabled).toBe(false);
    });
    fireEvent.press(button);
    await waitFor(() => {
      expect(mockSignup).toHaveBeenCalledWith("test@example.com", "validuser");
    });
  });
});
