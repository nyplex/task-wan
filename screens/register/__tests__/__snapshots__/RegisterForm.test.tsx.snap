// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm matches the snapshot 1`] = `
<View
  className="flex-col gap-2 mt-4"
>
  <View
    className="flex-col"
  >
    <View
      className="flex-row"
    >
      <View
        className="h-[48px] w-[50px] justify-center items-center bg-primary-0 rounded-l-[10px]"
      >
        <Text
          color="#fff"
          disabled={true}
          name="mail"
          size={24}
          testID="icon"
        >
          Feather Icon
        </Text>
      </View>
      <View
        className="flex-row overflow-hidden content-center data-[hover=true]:border-outline-400 data-[focus=true]:hover:border-primary-700 data-[disabled=true]:opacity-40 data-[disabled=true]:hover:border-background-300 items-center rounded border data-[invalid=true]:border-error-700 data-[invalid=true]:hover:border-error-700 data-[invalid=true]:data-[focus=true]:border-error-700 data-[invalid=true]:data-[focus=true]:hover:border-error-700 data-[invalid=true]:data-[disabled=true]:hover:border-error-700 data-[focus=true]:web:ring-1 data-[focus=true]:web:ring-inset data-[focus=true]:web:ring-indicator-primary data-[invalid=true]:web:ring-1 data-[invalid=true]:web:ring-inset data-[invalid=true]:web:ring-indicator-error data-[invalid=true]:data-[focus=true]:hover:web:ring-1 data-[invalid=true]:data-[focus=true]:hover:web:ring-inset data-[invalid=true]:data-[focus=true]:hover:web:ring-indicator-error data-[invalid=true]:data-[disabled=true]:hover:web:ring-1 data-[invalid=true]:data-[disabled=true]:hover:web:ring-inset data-[invalid=true]:data-[disabled=true]:hover:web:ring-indicator-error text-buttons-text flex-1 h-[48px] rounded-r-[10px] border-primary-700 data-[focus=true]:border-primary-0 rounded-l-none border-r-1 border-l-0"
        dataSet={
          {
            "disabled": "false",
            "focus": "false",
            "focusVisible": "false",
            "hover": "false",
            "invalid": "false",
            "readonly": "false",
            "required": "false",
          }
        }
        states={
          {
            "disabled": undefined,
            "focus": false,
            "focusVisible": false,
            "hover": undefined,
            "invalid": undefined,
            "readonly": undefined,
            "required": undefined,
          }
        }
        testID="input-field"
      >
        <TextInput
          accessible={true}
          aria-label="Input Field"
          aria-selected={false}
          autoCorrect={false}
          className="flex-1 text-typography-900 py-0 px-3 placeholder:text-typography-500 h-full ios:leading-[0px] web:cursor-text web:data-[disabled=true]:cursor-not-allowed web:outline-0 web:outline-none text-base"
          dataSet={
            {
              "disabled": "false",
              "focus": "false",
              "focusVisible": "false",
              "hover": "false",
              "invalid": "false",
              "readonly": "false",
              "required": "false",
            }
          }
          keyboardType="email-address"
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          onKeyPress={[Function]}
          placeholder="Email"
          placeholderTextColor="#9A9A9A"
          readOnly={false}
          secureTextEntry={false}
          states={
            {
              "disabled": undefined,
              "focus": false,
              "focusVisible": false,
              "hover": undefined,
              "invalid": undefined,
              "readonly": undefined,
              "required": undefined,
            }
          }
          type="text"
          value=""
        />
      </View>
    </View>
    <Text
      className="font-body based-styles font-medium text-[12px] tracking-[0px] text-right px-2 text-red-500 line-clamp-1"
    />
  </View>
  <View
    className="flex-col"
  >
    <View
      className="flex-row"
    >
      <View
        className="h-[48px] w-[50px] justify-center items-center bg-primary-0 rounded-l-[10px]"
      >
        <Text
          color="#fff"
          disabled={true}
          name="user"
          size={24}
          testID="icon"
        >
          Feather Icon
        </Text>
      </View>
      <View
        className="flex-row overflow-hidden content-center data-[hover=true]:border-outline-400 data-[focus=true]:hover:border-primary-700 data-[disabled=true]:opacity-40 data-[disabled=true]:hover:border-background-300 items-center rounded border data-[invalid=true]:border-error-700 data-[invalid=true]:hover:border-error-700 data-[invalid=true]:data-[focus=true]:border-error-700 data-[invalid=true]:data-[focus=true]:hover:border-error-700 data-[invalid=true]:data-[disabled=true]:hover:border-error-700 data-[focus=true]:web:ring-1 data-[focus=true]:web:ring-inset data-[focus=true]:web:ring-indicator-primary data-[invalid=true]:web:ring-1 data-[invalid=true]:web:ring-inset data-[invalid=true]:web:ring-indicator-error data-[invalid=true]:data-[focus=true]:hover:web:ring-1 data-[invalid=true]:data-[focus=true]:hover:web:ring-inset data-[invalid=true]:data-[focus=true]:hover:web:ring-indicator-error data-[invalid=true]:data-[disabled=true]:hover:web:ring-1 data-[invalid=true]:data-[disabled=true]:hover:web:ring-inset data-[invalid=true]:data-[disabled=true]:hover:web:ring-indicator-error text-buttons-text flex-1 h-[48px] rounded-r-[10px] border-primary-700 data-[focus=true]:border-primary-0 rounded-l-none border-r-1 border-l-0"
        dataSet={
          {
            "disabled": "false",
            "focus": "false",
            "focusVisible": "false",
            "hover": "false",
            "invalid": "false",
            "readonly": "false",
            "required": "false",
          }
        }
        states={
          {
            "disabled": undefined,
            "focus": false,
            "focusVisible": false,
            "hover": undefined,
            "invalid": undefined,
            "readonly": undefined,
            "required": undefined,
          }
        }
        testID="input-field"
      >
        <TextInput
          accessible={true}
          aria-label="Input Field"
          aria-selected={false}
          autoCorrect={false}
          className="flex-1 text-typography-900 py-0 px-3 placeholder:text-typography-500 h-full ios:leading-[0px] web:cursor-text web:data-[disabled=true]:cursor-not-allowed web:outline-0 web:outline-none text-base"
          dataSet={
            {
              "disabled": "false",
              "focus": "false",
              "focusVisible": "false",
              "hover": "false",
              "invalid": "false",
              "readonly": "false",
              "required": "false",
            }
          }
          keyboardType="default"
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          onKeyPress={[Function]}
          placeholder="Username"
          placeholderTextColor="#9A9A9A"
          readOnly={false}
          secureTextEntry={false}
          states={
            {
              "disabled": undefined,
              "focus": false,
              "focusVisible": false,
              "hover": undefined,
              "invalid": undefined,
              "readonly": undefined,
              "required": undefined,
            }
          }
          type="text"
          value=""
        />
      </View>
    </View>
    <Text
      className="font-body based-styles font-medium text-[12px] tracking-[0px] text-right px-2 text-red-500 line-clamp-1"
    />
  </View>
  <View
    className="mt-8"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      className="group/button flex-row items-center justify-center data-[focus-visible=true]:web:outline-none data-[focus-visible=true]:web:ring-2 data-[disabled=true]:opacity-40 gap-2 data-[hover=true]:bg-primary-600 border-primary-300 data-[hover=true]:border-primary-400 data-[active=true]:border-primary-500 data-[focus-visible=true]:web:ring-indicator-info px-5 based-styles w-full h-[48px] rounded-[10px] data-[active=true]:bg-primary-50 bg-[#C6C2C2]"
      collapsable={false}
      dataSet={
        {
          "active": "false",
          "disabled": "false",
          "focus": "false",
          "focusVisible": "false",
          "hover": "false",
        }
      }
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      role="button"
      states={
        {
          "active": false,
          "disabled": undefined,
          "focus": false,
          "focusVisible": false,
          "hover": false,
        }
      }
      testID="Button"
    >
      <View
        className="flex-row items-center justify-center gap-2 flex-1"
      >
        <Text
          className="based-styles font-regular text-[16px] tracking-[0px] text-typography-secondary"
        >
          Register
        </Text>
      </View>
    </View>
  </View>
</View>
`;
